react/features/transcription/middleware.js

```diff
@@ -0,0 +1,117 @@
+ */
+function _endpointMessageReceived({ dispatch, getState }, next, action) {
+    const payload = action.payload;
+
+    try {
+
+        // Let's first check if the given object has the correct
+        // topic in the payload, which identifies it as a json message send
+        // from Jigasi with speech-to-to-text results
+        // if (m['jitsi-meet-muc-msg-topic'] === TRANSCRIPTION_RESULT_TOPIC
+        //         && (payload = m.payload))
+        // TODO: To be replaced by if(payload['topic']==='transcription-result')
+        // after jigasi changes.
+        if (payload.transcript) {
+            // Extract the useful data from the payload of the JSON message
+            const isInterim = payload.is_interim;
+            const participantName = payload.participant.name;
+            const stability = payload.stability;
+            const text = payload.transcript[0].text;
+            const transcriptMessageID = payload.message_id;
+
+            // If this is the first result with the unique message ID,
+            // we add it to the state along with the name of the participant
+            // who said given text
+            if (!getState()['features/transcription']
+                .transcriptMessages.has(transcriptMessageID)) {
+                dispatch(addTranscriptMessage(transcriptMessageID,
+                    participantName));
+            }
+            const { transcriptMessages }
+                = getState()['features/transcription'];
+            const newTranscriptMessage
+                = transcriptMessages.get(transcriptMessageID);
+
+            // If this is final result, update the state as a final result
+            // and start a count down to remove the subtitle from the state
+            if (!isInterim) {
+
+                newTranscriptMessage.final = text;
+                dispatch(updateTranscriptMessage(transcriptMessageID,
+                    newTranscriptMessage));
+
+                setTimeout(() => {
+                    dispatch(removeTranscriptMessage(transcriptMessageID));
+                }, REMOVE_AFTER_MS);
+            } else if (stability > 0.85) {
+
+                // If the message has a high stability, we can update the
+                // stable field of the state and remove the previously
+                // unstable results
+
+                newTranscriptMessage.stable = text;
+                newTranscriptMessage.unstable = undefined;
+                dispatch(updateTranscriptMessage(transcriptMessageID,
+                    newTranscriptMessage));
+            } else {
+                // Otherwise, this result has an unstable result, which we
+                // add to the state. The unstable result will be localed at
+                // the end of the String, after the stable part.
+
+                newTranscriptMessage.unstable = text;
+                dispatch(updateTranscriptMessage(transcriptMessageID,
+                    newTranscriptMessage));
+            }
```

Comment:
I had a musing, and not a change request: what if instead of having add/update there was only one action that handled adding the new message if it doesn't exist or updating the existing message? Kind of like a POST request that creates the entity or updates the existing one; that way the double dispatch is avoided. Just a though that might trigger you to have other thoughts.

Reply:
As of now, the logic for updating the subtitles sits in the middleware entirely. If we use the same action for add/update we will need to shift some of it to the reducer. But will saving a dispatch make considerable improvements? If so, we can change it accordingly.

Reply:
I wouldn't say it's a considerable improvement, which is why I'm not suggesting that you actually make the change. It's just a little off to me that you do an explicit add and then next do an update when you can try to do them at the same time or do them as part of the add. So it's something I'm throwing out for thought only.